### ISSUE TITLE
Modularise jOOQ

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#6612]
We're already cross-releasing JDK 6/7 (commercial only) and JDK 8 (commercial and OSS) builds. In order to modularise jOOQ, we need to also add a new build for JDK 9.

Quite possibly, this will lead to some incompatible package renames, as we've re-used names like `org.jooq.util` in several modules in the past. These are:

- #7419

### Deployments

- The modular jOOQ will support JDK 11 (commercial and OSS, see #7789)
- The non-modular jOOQ will support JDK 6/7/8 (commercial only) and JDK 11 (commercial only)

### Regressions detected so far:

- [x] #10143 java.nio.file.InvalidPathException: Illegal char <*> when using FilePattern on module path
- [x] #10144 Runtime compilation error when compiling enum types on module path
- [x] #10145: NoClassDefFoundError when trying to load JAXB from a modularised jOOQ

----

Here's an alternative solution that might help avoiding cross releasing:
https://maven.apache.org/plugins/maven-compiler-plugin/examples/module-info.html

**Comment by lukaseder** [comment]
This is now implemented. There will be an ongoing set of fixes applied to this issue as problems are discovered, until jOOQ 3.12 is released.

This will also include updates to the manual.

**Comment by lukaseder** [comment]
The tool chain simply isn't ready. None of the IDEs manages to compile jOOQ correctly, making the development of jOOQ a major pain.

Modular jOOQ will be postponed to a future release

**Comment by cowwoc** [comment]
Hello from 2020. Is the tool chain support fixed yet?

**Comment by lukaseder** [comment]
Thanks for the ping. I haven't tried - but why not try again. The first impression is much better, i.e. I can add a `module-info.java` file and the IDE doesn't freeze instantly.

Let's see if this can be done in ~20 minutes.

**Comment by lukaseder** [comment]
What's your use-case for having jOOQ modularised?

**Comment by lukaseder** [comment]
Haha: 

> (package org.reactivestreams is declared in the unnamed module, but module org.reactivestreams does not read it)

**Comment by lukaseder** [comment]
Don't know what generated this. Now, all of a sudden, it's gone. Next step, getting the unit tests to work:
https://sormuras.github.io/blog/2018-09-11-testing-in-the-modular-world.html

**Comment by lukaseder** [comment]
Yeah, the usual can of worms, I guess JUnit needs to be upgraded to 5, finally. This is something I wanted to do anyway, eventually, now there's a reason (see also "Hal fixing light bulb").

But indeed, it looks like Eclipse and IntelliJ don't crash anymore

**Comment by lukaseder** [comment]
And just as I wrote this, Eclipse stopped working because I added a `src/test/java/module-info.java` file that opens the `org.jooq` module for testing.

Postponed for another 2 years.

**Comment by cowwoc** [comment]
> What's your use-case for having jOOQ modularised?

My application has multiple modules. One module exports a class that exposes JOOQ-specific types. This means that my module has to read "requires transitive org.jooq". The compiler then warns about "requires transitive" for automatic modules (if I recall correctly, it causes the entire classpath to get exported into the module path).

> And just as I wrote this, Eclipse stopped working because I added a src/test/java/module-info.java file that opens the org.jooq module for testing.

Can you please elaborate? What error did you get?

Also, do we really need all IDEs to support modules in order to add module support? Or is it a matter of you personally preferring to use Eclipse?

**Comment by lukaseder** [comment]
> The compiler then warns about "requires transitive" for automatic modules (if I recall correctly, it causes the entire classpath to get exported into the module path).

It complained about this as well in my case, when trying to `require transitive` org.reactivestreams`, which is an automatic module, not a real one. Then I removed transitive, added it again, *and it worked*! Not sure if that's good or bad.

But what's your use-case for having *your* application modularised?

> Can you please elaborate? What error did you get?

The project was displayed as empty in the project explorer. I.e. the kind of error that makes you doubt anything works at all.

> Also, do we really need all IDEs to support modules in order to add module support? Or is it a matter of you personally preferring to use Eclipse?

I work with Eclipse, yes. But I had had very similar experience with IntelliJ, last time I tried.

**Comment by cowwoc** [comment]
> But what's your use-case for having your application modularised?

I modularize all my code nowadays. For applications, you want to do this in order to use `jlink`. For my libraries, I modularize to avoid exporting implementation-specific details.

> I work with Eclipse, yes. But I had had very similar experience with IntelliJ, last time I tried.

Try IntelliJ 2020.1 (the latest stable version at this time). I haven't run into any problems with modules with it. The only gotcha to be aware of (not module-specific) is that they removed the Maven auto-import feature. Now, each time you modify the Maven configuration they display a tooltip inviting you to manually re-import.

Give IntelliJ a try and let me know how it goes.

**Comment by lukaseder** [comment]
I'll give this another try. We only have very few unit tests that test package-private stuff (what @sormuras calls whitebox testing in his excellent article). Most of our code is unit tested via public API (blackbox testing).

So, perhaps, I can get away with moving the whitebox tested stuff into some "Internal" (but exported) class, which would mean that we don't need the whole `module-info.[java|test]` thing.

I don't anyone working with jOOQ's code to be forced to using IDE X or Y, so a solution that doesn't work in Eclipse is a no-go.

> Now, each time you modify the Maven configuration they display a tooltip inviting you to manually re-import.

Welcome to my life, as an Eclipse developer :) (Or maybe, I turned auto-refreshing off...)

**Comment by lukaseder** [comment]
Note as it stands now, we're offering a Java 11 distribution only in the commercial editions, as there is currently only one distribution for the jOOQ Open Source Edition. We could increment its requirement to Java 11 (https://github.com/jOOQ/jOOQ/issues/7789), but I fear the market is not ready for this, yet.

Distributing more than one distribution of the jOOQ Open Source Edition is currently not on the roadmap.

**Comment by sormuras** [comment]
> ...  we don't need the whole `module-info.[java|test]` thing.

That's how I configured IDEA for internal tests at: https://github.com/sormuras/bach/tree/master/src/de.sormuras.bach

- The `test/java` folder contains the usual internal tests for "main" classes.
- The `test/java-module` folder contains only the `module-info.java` for build tools, that know how to handle it. That folder/file is ignored by IDE(A)s.

Works like a charm.

> Welcome to my life, as an Eclipse developer :) (Or maybe, I turned auto-refreshing off...)

/cc @sbrannen -- our Eclipse + Test + Modules expert in the JUnit Team, in case he has some Eclipse-related tricks to share. Like ignoring a file here or there.

**Comment by lukaseder** [comment]
> * The `test/java-module` folder contains only the `module-info.java` for build tools, that know how to handle it. That folder/file is ignored by IDE(A)s.

Sly hack 😁. Kinda reminds me of this hack here: https://twitter.com/lukaseder/status/1149256367619936261

**Comment by lukaseder** [comment]
> /cc @sbrannen -- our Eclipse + Test + Modules expert in the JUnit Team, in case he has some Eclipse-related tricks to share. Like ignoring a file here or there.

There's always a risk of a "regression", i.e. suddenly, the IDE "fixes" the ignoring of the file and things don't work anymore.

I'll try to avoid whitebox tests first. There are so few of them, and they're all avoidable.

**Comment by lukaseder** [comment]
We already have a variety of tests in a separate `jOOQ-test` module that were extracted from the `jOOQ` module because of some test instrumentation logic that depends on a `jooq.jar` file having been built. It seems that if I move all core unit tests into that separate module (which loads `jooq.jar` from the classpath again), then we can stay out of trouble, and even without any hacks!

The price to pay for this is a slightly slower feedback loop for things that can be unit tested (as opposed to integration tested), but I could live with that.

**Comment by lukaseder** [comment]
While I have everyone's attention: I personally doubt there is value in modularising any other modules than the core one. I remember that some of them could not be modularised because of the messy, split-packagy dependencies they have (e.g. `jOOQ-checker` because of the checker framework, or `jOOQ-meta-extensions` because of hibernate).

Thoughts?

**Comment by lukaseder** [comment]
I don't see any issues yet. We'll ship this in jOOQ 3.14 for the commercial editions' Java 11 distribution.

The only remaining question is what to do with the jOOQ Open Source Edition: https://github.com/jOOQ/jOOQ/issues/7789

Options are:

- Keep it at Java 8 for now (then this issue here does not apply yet)
- Upgrade it to Java 11 (bold move, but why not)
- Release more than 1 distribution (I'd personally like to keep this a paid service)

**Comment by cowwoc** [comment]
@lukaseder If you want to kill two birds with one stone (and hey, why not) I would force the open-source version to use newer JDK versions and only support older JDKs for the commercial version.

In my experience, enterprise customers are stuck on JDK 8. They expect and are willing to pay extra for supporting older JDKs.

**Comment by cowwoc** [comment]
> While I have everyone's attention: I personally doubt there is value in modularising any other modules than the core one.

I only use the core module so I'm fine stopping at this point but maybe other users will disagree.

**Comment by lukaseder** [comment]
> I only use the core module so I'm fine stopping at this point but maybe other users will disagree.

Let's collect traction for this in a separate issue: #10149

**Comment by lukaseder** [comment]
Fixing the edition labels on this issue. I've decided that Java 8 still enjoys too much adoption in the market to offer a modular jOOQ Open Source Edition, forcing everyone to upgrade or pay for this minor feature here.

**Comment by cowwoc** [comment]
@lukaseder We can have our cake and eat it too. Using the multi-release JAR feature, you can configure the artifact to provide Java Module support if it is enabled (Java 9+ users) and ignore it for older versions (Java 8 users). I recently contributed a patch to HikariCP that did the same and it works great: https://github.com/brettwooldridge/HikariCP/issues/1572

This consists of two simple steps:

1. Add `META-INF/versions/9/module-info.class` to the JAR.
2. Add `Multi-Release: true` to `META-INF/MANIFEST.MF`.

If I contribute a PR for this, would you consider adding JPMS support to version 3.14.1?

**Comment by lukaseder** [comment]
Nice hack, @cowwoc. So, you're using ant to build just this `module-info.class` file and place it in the Java 8 deliverable? And the Java 8 runtime will ignore it, because it doesn't know the file layout of `META-INF/version9/...`?

I'll play around with this. I think that'll be easier than a PR, because you don't have access to the "real" build. But if that works, then yes, this can be included in 3.14.1.

**Comment by lukaseder** [comment]
Nah, doesn't work. A new set of errors starts popping up, and after another 1h on this, I have spent more than enough time on modularity.

The commercial editions have 3 distributions: Java 11, 8, 6. The one built for Java 11+ is modular. As soon as we bump the jOOQ Open Source Edition's required JDK to 11, it will be modular as well.

**Comment by cowwoc** [comment]
@lukaseder I'm sorry ran into problems. From the sound of it, you got tripped up on the very first step. You can't build `module-info.class` in isolation because you'll run into precisely the kind of problems you've mentioned. Let me help you out. I'll publish a PR for you to review.

I assume you have automated builds that run against submitted PRs? Do you have an existing branch I should jump on to pick up your existing changes?

**Comment by lukaseder** [comment]
I appreciate your offers, and your persistence. I know this feature means a lot to you.

I won't keep you from creating a PR, and possibly building your own distribution. Or even better, you use the jOOQ Professional Edition, which is ready for you, plus a ton of great premium features and access to commercial RDBMS!

It's not just this attempt. Everytime I touch JPMS, I waste several hours. Things work now the way jOOQ is built in editions and distributions. Hardly anyone else has requested this feature, and whenever I made twitter polls, people were very surprised that I even try. There are far more interesting things on the roadmap that I'd like to spend my time on.

> I assume you have automated builds that run against submitted PRs?

No, we don't, because we hardly accept PRs, because usually, reviewing a PR is more work than doing it myself, and I don't like the idea that anyone can submit work and thus reprioritise the product roadmap. Unfortunately, the feature cannot be turned off on github without turning off the issue tracker as well.

You don't have the entire Jenkins pipeline available with the currently 17 different builds on currently 5 maintained branches. Or the tests, including the ones that check if jOOQ for Java 11 can be correctly used on the classpath and on the module path. Nor do you have access to my Eclipse environment, which tripped up again today with the alternative setup of having the `module-info.java` elsewhere (didn't even check IntelliJ). Also, the module path tests kept failing today when I tried this (the classpath tests worked). Perhaps yet another set of mistakes I made, but again, potential hours wasted if I investigate this.

Remember how I had to move all the tests out of the main module? https://github.com/jOOQ/jOOQ/issues/6612#issuecomment-621858216 Do you know how many unnecessary conflicts this produced whenever I backported an issue to older branches? Or the lost history on those test files, because the only thing I wasted more time on in my life other than JPMS is git?

I'm *done* with JPMS. *done* :) It works now. (Thanks for listening).

**Comment by cowwoc** [comment]
Sounds stressful. I'm sorry you had a bad experience.

If you change your mind in the future and you are open to collaborating on this, please let me know.

In the meantime, you spoke of bumping the Open Source edition's required JDK to 11. Do you have a timetable for that?

**Comment by lukaseder** [comment]
Though I made it sound stressful, this is about cost/benefit. I don't think I'll change my mind, unless this issue gets a lot more traction (benefit). Compare it to this issue, for example, the currently highest upvoted issue: https://github.com/jOOQ/jOOQ/issues/6298 (also very high cost, but high benefit, too).

I wish I could push the JDK requirements further, but it's just not reasonable at this time. It looks like the market isn't ready. Look at these numbers:
https://twitter.com/vlad_mihalcea/status/1302516212887322624

![image](https://user-images.githubusercontent.com/734593/96963273-abbacd80-1508-11eb-8c93-1bf3ee8bb99c.png)

Things will change with the next LTS, I'm sure. Project Amber adds *a lot* of value so the upgrade will be worth it. JPMS + the 2-3 syntax sugars of Java 11 simply aren't a good enough incentive for folks to upgrade (and pay the high price of fixing 100s of JAXB related issues, for example).

I see, you've edited your recent comment:

> PS: If all else fails, we can simply steal `module-info.class` from the JDK 11 commercial edition. It would look something like this

In our build, which is currently building all editions and distributions in parallel and completely independently from one another, we'd have to make significant efforts to re-order stuff to introduce such a dependency. That's another yak not worth shaving. The ant hack sounded more promising, at first. But it's a hack, and it didn't work. So does the Maven compiler thing. But we're building the Java 8 distribution with Java 8 not with Java 11. Changing that will lead to another set of things that need to be adapted. Do you know [Malcolm in the middle](https://www.youtube.com/watch?v=AbSehcT19u0)?

The jOOQ Professional Edition (or even the Express Edition if this is about personal projects) has a modular JDK 11 distribution that is integration tested and works just fine. Given the popularity of this feature request, this is a power user feature for now. Given the time spent discussing it (I don't know your hourly rate), you would have gotten a cheaper solution by now, by just upgrading to that distribution :)

**Comment by cowwoc** [comment]
That's Malcolm in the middle video clip is hilarious :) Thanks for the good laugh.

**Comment by lukaseder** [comment]
https://github.com/jOOQ/jOOQ/issues/7789 is integrated in jOOQ 3.15.0, meaning that starting with 3.15, the minimum JDK requirement of the jOOQ Open Source Edition is Java 11 and it automatically turns modular.
